src/librustc_mir/borrow_check/move_errors.rs

```diff
@@ -0,0 +1,364 @@
+
+pub(crate) fn report_move_errors<'gcx, 'tcx>(
+    mir: &Mir<'tcx>,
+    tcx: TyCtxt<'_, 'gcx, 'tcx>,
+    move_errors: Vec<MoveError<'tcx>>,
+    move_data: &MoveData<'tcx>,
+) {
+    MoveErrorCtxt {
+        mir,
+        tcx,
+        move_data,
+    }.report_errors(move_errors);
+}
+
+#[derive(Copy, Clone)]
+struct MoveErrorCtxt<'a, 'gcx: 'tcx, 'tcx: 'a> {
+    mir: &'a Mir<'tcx>,
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    move_data: &'a MoveData<'tcx>,
+}
+
+#[derive(Debug)]
+enum GroupedMoveError<'tcx> {
+    // Match place can't be moved from
+    // e.g. Matching on x[0].
+    MovesFromMatchPlace {
+        span: Span,
+        move_from: Place<'tcx>,
+        kind: IllegalMoveOriginKind<'tcx>,
+        binds_to: Vec<Local>,
+    },
+    // Part of a pattern can't be moved from,
+    // e.g. moving from a &x pattern.
+    MovesFromPattern {
+        span: Span,
+        move_from: MovePathIndex,
+        kind: IllegalMoveOriginKind<'tcx>,
+        binds_to: Vec<Local>,
+    },
+    OtherIllegalMove {
+        span: Span,
+        kind: IllegalMoveOriginKind<'tcx>,
+    },
+}
+
+impl<'a, 'gcx, 'tcx> MoveErrorCtxt<'a, 'gcx, 'tcx> {
+    fn report_errors(self, move_errors: Vec<MoveError<'tcx>>) {
+        let grouped_errors = self.group_move_errors(move_errors);
+        for error in grouped_errors {
+            self.report(error);
+        }
+    }
+
+    fn group_move_errors(self, errors: Vec<MoveError<'tcx>>) -> Vec<GroupedMoveError<'tcx>> {
+        let mut grouped_errors = Vec::new();
+        for error in errors {
+            self.append_to_grouped_errors(&mut grouped_errors, error);
+        }
+        grouped_errors
+    }
+
+    fn append_to_grouped_errors(
+        self,
+        grouped_errors: &mut Vec<GroupedMoveError<'tcx>>,
+        error: MoveError<'tcx>,
+    ) {
+        match error {
+            MoveError::UnionMove { .. } => {
+                unimplemented!("don't know how to report union move errors yet.")
+            }
+            MoveError::IllegalMove {
+                cannot_move_out_of: IllegalMoveOrigin { location, kind },
+            } => {
+                let stmt_source_info = self.mir.source_info(location);
+                if let Some(StatementKind::Assign(
+                    Place::Local(local),
+                    Rvalue::Use(Operand::Move(move_from)),
+                )) = self.mir.basic_blocks()[location.block]
+                    .statements
+                    .get(location.statement_index)
+                    .map(|stmt| &stmt.kind)
+                {
+                    let local_decl = &self.mir.local_decls[*local];
+                    if let Some(ClearCrossCrate::Set(BindingForm::Var(VarBindingForm {
+                        ref opt_match_place,
+                        binding_mode: _,
+                        opt_ty_info: _,
+                    }))) = local_decl.is_user_variable
+                    {
+                        let (match_place, match_span) = opt_match_place
```

Comment:
I'd like to see some sort of comment explaining what these represent in terms of Rust source (e.g., give a Rust example, and explain what `match_place` and `match_span` would be)